### PR TITLE
Add Crop Type Mapping Tutorial

### PR DESCRIPTION
**Pull request by yichiac**
This PR provides a tutorial for using spatiotemporal segmentation task trainer #3671 to train a ConvLSTM model for crop type mapping using the PASTIS dataset.

We will continue the tutorial of training on PASTIS and inferencing on Germany using EuroCrops. We can demo the new sampler #3552 during the inference stage. 

This PR also fixes the normalization in PASTIS datamodule. The default normalization for NonGeoDataset is mean=0, std=255; PASTIS normalization should be mean=0, std=10000. 

Part of #2382 
Closes #2449

@robmarkcole @adamjstewart 

## AI Usage Disclosure

- [ ] 🟢 **No AI usage**: written by humans, for humans
- [x] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by robmarkcole** [comment]
@yichiac I made a few changes to https://github.com/torchgeo/torchgeo/pull/3671 on Friday so make sure you pull those changes. @adamjstewart perhaps we can get that over the line so we can discuss this PR on Weds?

**Comment by adamjstewart** [comment]
FYI, https://github.com/torchgeo/torchgeo/pull/3671 is finally merged!

File: torchgeo/datamodules/pastis.py [hunk 1]
```diff
@@ -21,6 +21,9 @@ class PASTISDataModule(NonGeoDataModule):
     .. versionadded:: 0.8
     """
 
+    mean = torch.tensor(0)
```
Comment (robmarkcole) [inline torchgeo/datamodules/pastis.py:24]: Should these be floats?

**Comment by robmarkcole** [comment]
@adamjstewart I've created https://github.com/torchgeo/torchgeo/pull/3715 I think this paper would be even stronger with > 1 model

**Comment by robmarkcole** [comment]
@yichiac 
image

**Comment by adamjstewart** [comment]
> @adamjstewart I've created #3715 I think this paper would be even stronger with > 1 model

If it was hand-written, sure. As is, this might take another month to review/merge. Would much rather get OlmoEarth working since it's already merged.

File: torchgeo/datamodules/pastis.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datamodules/pastis.py:24]: PyTorch automatically casts ints to floats when necessary.

**Comment by robmarkcole** [comment]
Olmo is pending https://github.com/allenai/olmoearth_pretrain/pull/559 I don't see it getting merged before the deadline

**Comment by robmarkcole** [comment]
@adamjstewart alternatively we might be able to get https://github.com/torchgeo/torchgeo/pull/3086 and a regression trainer done in time, WDYT? My own use is for regression so quite motivated

**Comment by adamjstewart** [comment]
Happy to get that merged but don't think we have room for more than 4 tutorials in the paper.

**Comment by robmarkcole** [comment]
Created https://github.com/torchgeo/torchgeo/pull/3727

**Comment by yichiac** [comment]
> @yichiac image

Oops, sorry about that. It seems that the metadata caused the issue. It has been fixed.

File: docs/tutorials/case_studies.rst [hunk 2]
```diff
@@ -7,6 +7,7 @@ The following case studies present end-to-end workflows for common use cases of
 * `Earth Surface Water <earth_surface_water.ipynb>`_: A workflow for mapping surface water, including lakes and rivers
 * `Earthquake Detection <earthquake_detection.ipynb>`_: A workflow for change detection to detect areas affected by earthquakes
 * `NAIP Road Segmentation <naip_road_segmentation.ipynb>`_: A workflow for road extraction from NAIP imagery using pre-trained segmentation weights
+* `Spatiotemporal Segmentation for Crop Type Mapping <spatiotemporal_segmentation.ipynb>`_: A workflow for time-series semantic segmentation to identify crop type mapping
```
Comment (adamjstewart) [inline docs/tutorials/case_studies.rst]: Can we change the link, filename, and title to "Crop Type Mapping" for consistency? We will likely add other spatiotemporal segmentation tutorials in the future, crop type mapping is just one specific example.

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 3]
```diff
@@ -0,0 +1,968 @@
...
+    "import matplotlib.pyplot as plt\n",
+    "import torch\n",
+    "from lightning.pytorch import Trainer\n",
+    "from lightning.pytorch.callbacks import EarlyStopping, ModelCheckpoint\n",
+    "\n",
+    "from torchgeo.datamodules import PASTIS100DataModule\n",
+    "from torchgeo.datasets import PASTIS100\n",
+    "from torchgeo.trainers import SpatioTemporalSegmentationTask\n",
+    "\n",
+    "torch.set_float32_matmul_precision('medium')\n",
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: Is this required? It hasn't been explained in the tutorial. Readers of the paper will have no idea what this means or why it doesn't work on XYZ hardware.

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 4]
```diff
@@ -0,0 +1,968 @@
...
+   "id": "7",
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 404
+    },
+    "id": "7",
+    "outputId": "3adc36d2-8371-4010-8f9f-7338398180b9"
+   },
+   "outputs": [
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: Let's strip all outputs for GitHub, otherwise the diff on this file is massive every time a change is made. Also, the visualization could be greatly improved with:

1. A GIF for time-series imagery
2. A legend for semantic segmentation classes

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 5]
```diff
@@ -0,0 +1,968 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "6de57690",
+   "metadata": {
+    "id": "6de57690"
+   },
+   "source": [
+    "## Datamodule\n",
+    "\n",
+    "We will utilize TorchGeo's [Lightning](https://lightning.ai/docs/pytorch/stable/) datamodules to organize the dataloader setup."
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: This doesn't really tell me the important details like train-val-test split, normalization, data augmentation, etc. The paper won't just be a list of code, it needs detailed paragraphs explaining each step and what is happening behind the scenes so readers can understand how to modify things or what the consequences of these decisions are.

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 6]
```diff
@@ -0,0 +1,968 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "70b326ea",
+   "metadata": {
+    "id": "70b326ea"
+   },
+   "source": [
+    "## Spatio-temporal Segmentation Task\n",
+    "\n",
+    "`SpatioTemporalSegmentationTask` is a trainer designed for time-series imagery with static segmentation labels."
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: Needs a detailed description, citation, and motivation for ConvLSTM, DICE loss, and all other hyperparameter choices. What does this hyperparam control, and how should readers go about performing hyperparam tuning? What values typically work well, and why were these specific params chosen?

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 7]
```diff
@@ -0,0 +1,968 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "74e98a4d",
+   "metadata": {
+    "id": "74e98a4d"
+   },
+   "source": [
+    "## Training a deep learning model from scratch\n",
+    "\n",
+    "We can use Lightning's `Trainer` and TorchGeo's `SpatioTemporalSegmentationTask` to simplify the training."
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: What is early stopping? What is model checkpointing? Why should readers use these? What does `.fit()` do?

File: docs/tutorials/spatiotemporal_segmentation.ipynb [hunk 8]
```diff
@@ -0,0 +1,968 @@
...
+       "  'test_AverageJaccardIndex': 0.33454036712646484,\n",
+       "  'test_AveragePrecision': 0.4086105227470398,\n",
+       "  'test_AverageRecall': 0.4434086084365845,\n",
+       "  'test_ClasswiseF1Score_0': 0.692541241645813,\n",
+       "  'test_ClasswiseF1Score_1': 0.66221022605896,\n",
+       "  'test_ClasswiseF1Score_2': 0.8551112413406372,\n",
+       "  'test_ClasswiseF1Score_3': 0.7552766799926758,\n",
+       "  'test_ClasswiseF1Score_4': 0.7352511882781982,\n",
+       "  'test_ClasswiseF1Score_5': 0.8869552612304688,\n",
+       "
```
Comment (adamjstewart) [inline docs/tutorials/spatiotemporal_segmentation.ipynb]: Can this entire section be replaced with a single-line tensorboard call? We log all prediction plots to tensorboard, why not use them?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Does not yet look like a magazine article/paper.

We can keep the notebook file using PASTIS100 but feel free to switch to PASTIS for the actual paper.

File: docs/tutorials/spatiotemporal_segmentation.ipynb (hunk 4 above)
Comment (yichiac) [reply docs/tutorials/spatiotemporal_segmentation.ipynb]: Outputs stripped. We can leave the GIF and legend for the next PR.

File: docs/tutorials/spatiotemporal_segmentation.ipynb (hunk 8 above)
Comment (yichiac) [reply docs/tutorials/spatiotemporal_segmentation.ipynb]: The current `SpatioTemporalSegmentationTask` does not include the `add_figure` method to log prediction plots, unlike `SemanticSegmentationTask`. We can port the plotting block from `SemanticSegmentationTask.validation_step` into `SpatioTemporalSegmentationTask` in a future PR.

**Comment by yichiac** [comment]
Added a block to resolve the notebook check. I don't like adding it just for notebook checks. Not sure if we can bypass the checks without adding `fast_dev_run`.

**Comment by adamjstewart** [comment]
Let me know when you've incorporated all of the changes from the reviewers, then I'll review again.

**Comment by adamjstewart** [comment]
> Not sure if we can bypass the checks without adding fast_dev_run.

You can add the tutorial to the list in `docs/tutorials/conftest.py`

**Comment by robmarkcole** [comment]
I would be curious how the results are with Utae which tends to be one of the models people compare against

**Comment by yichiac** [comment]
@robmarkcole thanks! Once #3715 is merged, I can add results using utae.

**Comment by robmarkcole** [comment]
@adamjstewart any thoughts on moving utae along? It's been stalled for a while now

**Comment by robmarkcole** [comment]
@yichiac utae is merged
